File: source/parlex/src/wirth.cpp
#include "parlex/detail/wirth.hpp"

#include "parlex/detail/builtins.hpp"

#include "parlex/detail/parser.hpp"

#include "utils.hpp"
#include "parlex/builder.hpp"
#include "utf.hpp"
#include <iostream>
#include "parlex/detail/c_string.hpp"

namespace parlex {
namespace detail {

erased<node> wirth_t::process_factor(std::u32string const & document, ast_node const & factor) const {

	auto i = factor.children.begin();
	std::string tag;
	if (i->recognizer_index == dollar_sign) {
		++i;
		if (i->recognizer_index == identifier_dfa) {
			tag = tolower(to_utf8(document.substr(i->document_position, i->consumed_character_count)));
		} else if (i->recognizer_index == c_string_recognizer_index) {
			tag = to_utf8(c_string_t::extract(*this, document, *i));
		} else {
			throw;
		}
	}
	
	if (i->recognizer_index == tag_dfa) {
		auto const & j = i->children[1];
		++i;
		throw_assert(j.recognizer_index == identifier_dfa);
		tag = to_utf8(document.substr(j.document_position, j.consumed_character_count));
	}

	std::unique_ptr<erased<node>> result;
	auto const set = [&result](erased<node> const & node) { result.reset(new erased<detail::node>(node)); };
	if (i->recognizer_index == identifier_dfa) {
		auto const name = to_utf8(document.substr(i->document_position, i->consumed_character_count));
		set(reference(name));
	} else if (i->recognizer_index == c_string_recognizer_index) {
		auto const text = c_string_t::extract(*this, document, *i);
		set(literal(text));
	} else if (i->recognizer_index == parenthetical_dfa) {
		auto const & q = i->children;
		auto j = q.begin();
		auto const parentheticalType = j->recognizer_index;
		while (j->recognizer_index != expression_dfa) {
			++j;
		}
		if (parentheticalType == open_square) {
			set(optional(process_expression(document, *j)));
		} else if (parentheticalType == open_curly) {
			set(repetition(process_expression(document, *j)));
		} else if (parentheticalType == open_paren) {
			set(process_expression(document, *j));
		} else {
			throw;
		}
	} else {
		throw;
	}

	if (!tag.empty()) {
		(*result)->tag = tag;
	}
	return *result;
}

erased<node> wirth_t::process_term(std::u32string const & document, ast_node const & term) const {
	std::vector<ast_node const *> factors;

	for (auto const & entry : term.children) {
		if (entry.recognizer_index == factor_dfa) {
			factors.push_back(&entry);
		}
	}
	if (factors.size() == 1) {
		return process_factor(document, *factors[0]);
	}
	sequence result;
	for (auto factor : factors) {
		result.children.push_back(process_factor(document, *factor));
	}
	return result;
}

erased<node> wirth_t::process_expression(std::u32string const & document, ast_node const & expression) const {
	std::vector<ast_node const *> terms;

	auto const & p = expression.children;
	for (auto const & entry : p) {
		if (entry.recognizer_index == term_dfa) {
			terms.push_back(&entry);
		}
	}
	if (terms.size() == 1) {
		return process_term(document, *terms[0]);
	}
	choice result;
	for (auto term : terms) {
		result.children.push_back(process_term(document, *term));
	}
	return result;
}

erased<node> wirth_t::compile_expression(std::u32string const & source) const {
	parser p;
	auto assl = p.parse(*this, get_recognizer(expression_dfa), source, progress_bar);
	if (!assl.is_rooted()) {
		throw std::runtime_error("could not parse expression");
	}
	auto const ast = assl.tree();
	return process_expression(source, ast);
}

builder wirth_t::load_grammar(std::string const & rootId, std::map<std::string, production> const & definitions) const {
	std::set<std::string> names;
	builder result;
	result.root_name = rootId;
	for (auto const & entry : definitions) {
		auto const & id = entry.first;
		auto const & definition = entry.second;
		auto const res = names.insert(entry.first);
		if (!res.second) {
			throw std::runtime_error(("duplicate reference ID " + id).c_str());
		}
		result.productions.emplace_back(id, compile_expression(definition.source), definition.assoc, definition.filter, definition.precedences);
	}
	return result;
}


builder wirth_t::load_grammar(std::string const & rootId, std::u32string const & document, std::map<std::string, associativity> const & associativities, std::set<std::string> const & longestNames, std::set<std::string> const & shortestNames) const {
	parser p;
	auto assl = p.parse(*this, document);
	auto const ast = assl.tree();
	std::set<std::u32string> names;
	std::map<std::string, production> definitions;
	auto const & top = ast.children;
	for (auto const & entry : top) {
		if (entry.recognizer_index == production_dfa) {
			auto const & namePart = entry.children[0];
			auto const u32Name = document.substr(namePart.document_position, namePart.consumed_character_count);
			auto const name = to_utf8(u32Name);
			auto const res = names.insert(u32Name);
			if (!res.second) {
				throw std::runtime_error(("duplicate reference ID " + name).c_str());
			}
			std::u32string source;
			for (auto const & entry2 : entry.children) {
				if (entry2.recognizer_index == expression_dfa) {
					source = document.substr(entry2.document_position, entry2.consumed_character_count);
					break;
				}
			}
			auto assoc = associativities.count(name) > 0 ? associativities.find(name)->second : associativity::NONE;
			auto filter = filter_function();
			if (longestNames.count(name) > 0) {
				filter = longest();
			} else if (shortestNames.count(name) > 0) {
				filter = shortest();
			}
			definitions.emplace(std::piecewise_construct, forward_as_tuple(name), forward_as_tuple(source, assoc, filter, std::set<std::string>()));
		}
	}
	return load_grammar(rootId, definitions);
}

wirth_t::production::production(std::u32string const & source, associativity const assoc, filter_function const & filter, std::set<std::string> const & precedences) : source(source), assoc(assoc), filter(filter), precedences(precedences) {
}

builder generate_wirth() {
	return builder("root", {
		production("whiteSpace",
			sequence({
				reference("white_space"),
				repetition(reference("white_space"))
			}),
			associativity::NONE, longest()
		),

		production("comment",
			sequence({
				literal(U"#"),
				repetition(reference("not_newline")),
				literal(U"\n")
			})
		),

		production("identifier",
			sequence({
				choice({
					reference("letter"),
					literal(U"_")
				}),
				repetition(
					choice({
						reference("letter"),
						literal(U"_"),
						reference("decimal_digit")
					})
				)
			}),
			associativity::NONE, longest()
		),

		production("production", 
			sequence({
				reference("identifier"),
				optional(reference("whiteSpace")),
				literal(U"="),
				optional(reference("whiteSpace")),
				reference("expression"),
				optional(reference("whiteSpace")),
				literal(U".")
			})
		),

		production("expression", 
			sequence({
				reference("term"),
				repetition(
					sequence({
						optional(reference("whiteSpace")),
						literal(U"|"),
						optional(reference("whiteSpace")),
						reference("term")
					})
				)
			})
		),

		production("term", 
			sequence({
				reference("factor"),
				repetition(
					sequence({
						optional(reference("whiteSpace")),
						reference("factor")
					})
				)
			})
		),

		production("parenthetical",
			choice({
				sequence({
					literal(U"["),
					optional(reference("whiteSpace")),
					reference("expression"),
					optional(reference("whiteSpace")),
					literal(U"]")
				}),
				sequence({
					literal(U"("),
					optional(reference("whiteSpace")),
					reference("expression"),
					optional(reference("whiteSpace")),
					literal(U")")
				}),
				sequence({
					literal(U"{"),
					optional(reference("whiteSpace")),
					reference("expression"),
					optional(reference("whiteSpace")),
					literal(U"}")
				})
			})
		),

		production("tag",
			sequence({
				literal(U"%"),
				reference("identifier")
			})
		),

		production("factor", 
			choice({
				sequence({
					optional(literal(U"$")),
					reference("identifier")
				}),
				sequence({
					optional(literal(U"$")),
					reference("c_string")
				}),
				sequence({
					optional(reference("tag")),
					reference("parenthetical")
				})
			}),
			associativity::NONE, filter_function(), {"factor"}
		),

		production("root", 
			repetition(choice({
				reference("production"),
				reference("comment"),
				reference("whiteSpace")
			}))
		)
	}
);

}

wirth_t::wirth_t() : grammar(generate_wirth()),
                     open_square(lookup_literal_recognizer_index(U"[")),
                     open_paren(lookup_literal_recognizer_index(U"(")),
                     open_curly(lookup_literal_recognizer_index(U"{")),
                     dollar_sign(lookup_literal_recognizer_index(U"$")),

                     production_dfa(lookup_recognizer_index("production")),
                     expression_dfa(lookup_recognizer_index("expression")),
                     term_dfa(lookup_recognizer_index("term")),
                     parenthetical_dfa(lookup_recognizer_index("parenthetical")),
                     tag_dfa(lookup_recognizer_index("tag")),
                     factor_dfa(lookup_recognizer_index("factor")),
                     identifier_dfa(lookup_recognizer_index("identifier")),
					 c_string_recognizer_index(grammar::lookup_recognizer_index("c_string")) {}

erased<node> wirth_t::process_production(std::u32string const & document, ast_node const & production) const {
	for (auto const & entry : production.children) {
		if (entry.recognizer_index == expression_dfa) {
			return process_expression(document, entry);
		}
	}
	throw;
}


} //namespace detail
detail::wirth_t const& wirth() {
	static detail::wirth_t value;
	return value;
}
} //namespace parlex

File: source/python_tools/wirth.py
class Option:
    def __init__(self, child):
        self.child = child

class Repetition:
    def __init__(self, child):
        self.child = child

class Sequence:
    def __init__(self, children):
        self.children = children

class Variant:
    def __init__(self, children):
        self.children = children

class Identifier:
    def __init__(self, name):
        self.name = name

class Literal:
    def __init__(self, text):
        self.text = text

class Production:
    def __init__(self, name, text):
        self.name = name
        (i, self.tree) = self.parse_expression(text, 0)

    def skip_whitespace(self, text, i):
        while i < len(text) and text[i].isspace():
            i += 1
        return i

    def parse_expression(self, text, i):
        old_i = i
        children = list()
        i = self.skip_whitespace(text, i)
        while True:
            (i, term) = self.parse_term(text, i)
            i = self.skip_whitespace(text, i)
            if term:
                children.append(term)
            if not term or i >= len(text) or not text[i] == "|":
                break
            i += 1
        if len(children) == 0:
            return (old_i, None)
        elif len(children) == 1:
            return (i, children[0])
        else:
            return (i, Variant(children))

    def parse_term(self, text, i):
        old_i = i
        children = list()
        i = self.skip_whitespace(text, i)
        while True:
            (i, factor) = self.parse_factor(text, i)
            i = self.skip_whitespace(text, i)
            if factor:
                children.append(factor)
            if not factor or i >= len(text):
                break
        if len(children) == 0:
            return (old_i, None)
        elif len(children) == 1:
            return (i, children[0])
        else:
            return (i, Sequence(children))

    def parse_factor(self, text, i):
        (i, identifier) = self.parse_identifier(text, i)
        if identifier:
            return (i, identifier)
        (i, literal) = self.parse_literal(text, i)
        if literal:
            return (i, literal)

        old_i = i
        i = self.skip_whitespace(text, i)
        if i < len(text):
            if text[i] == "[":
                i += 1
                (i, expression) = self.parse_expression(text, i)
                if expression:
                    i = self.skip_whitespace(text, i)
                    if not i < len(text) or not text[i] == "]":
                        raise Exception("expected ] at " + str(i))
                    i += 1
                    return (i, Option(expression))

            if text[i] == "(":
                i += 1
                (i, expression) = self.parse_expression(text, i)
                if expression:
                    i = self.skip_whitespace(text, i)
                    if not i < len(text) or not text[i] == ")":
                        raise Exception("expected ) at " + str(i))
                    i += 1
                    return (i, expression)

            if text[i] == "{":
                i += 1
                (i, expression) = self.parse_expression(text, i)
                if expression:
                    i = self.skip_whitespace(text, i)
                    if not i < len(text) or not text[i] == "}":
                        raise Exception("expected } at " + str(i))
                    i += 1
                    return (i, Repetition(expression))

        return (old_i, None)

    def parse_identifier(self, text, i):
        old_i = i
        i = self.skip_whitespace(text, i)
        result = ""
        if i < len(text):
            if text[i].isalpha():
                result += text[i]
                i += 1
                while i < len(text) and (text[i].isalpha() or text[i] == "_"):
                    result += text[i]
                    i += 1
        if not result == "":
            return (i, Identifier(result))
        else:
            return (old_i, None)


    def parse_literal(self, text, i):
        escapes = {
            "n": "\n",
            "r": "\r",
            "t": "\t",
            "b": "\b",
            "f": "\f",
            "v": "\v",
            "0": "\0"
        }

        old_i = i
        i = self.skip_whitespace(text, i)
        if i >= len(text) or text[i] != "\"":
            return (old_i, None)
        i += 1
        result = ""
        while i < len(text) and not text[i] == "\"":
            if text[i] == "\\":
                i += 1
                if i < len(text):
                    c = text[i]
                    i += 1
                    if c in escapes:
                        result += escapes[c]
                    else:
                        result += c
            else:
                result += text[i]
                i += 1
        if not text[i] == "\"":
            raise Exception("expected \" at " + str(i))
        i += 1
        return (i, Literal(result))